Edits are processed in the order of submission

hubertp opened the pull request:
### Pull Request Description

This is a quick fix to a long standing problem of
`org.enso.interpreter.service.error.FailedToApplyEditsException` which would prevent backend from processing any more changes, rendering GUI (and backend) virtually useless.
Edits are submitted for (background) processing in the order they are handled. However the order of execution of such tasks is not guaranteed. Most of the time edits are processed in the same order as their requests but when they don't, files get quickly out of sync.

Related to #8770.

### Important Notes

I'm not a fan of this change because it essentially blocks all open/file requests until all edits are processed and we already have logic to deal with that appropriately. Moreover those tasks can and should be processed independently. Since we already had the single thread executor present to ensure correct synchronization of open/file/push commands, we are simply adding edit commands to the list.

Ideally we want to have a specialized executor that executes tasks within the same group sequentially but groups of tasks can be executed in parallel, thus ensuring sufficient throughput. The latter will take much longer and will require significant rewrite of the command execution.

Added tests that would previously fail due to non-deterministic execution.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Edits are processed in the order of submission

This is a quick fix to a long standing problem of
`org.enso.interpreter.service.error.FailedToApplyEditsException` which
would prevent backend from processing any more changes, rendering GUI
(and backend) virtually useless.
Edits are submitted for (background) processing in the order they are
handled. However the order of execution of such tasks is not guaranteed.
Most of the time edits are processed in the same order as their requests
but when they don't, files get quickly out of sync.

I'm not a fan of this change because it essentially blocks all open/file
requests until all edits are processed and we already have logic to deal
with that appropriately. Moreover those tasks can and should be
processed independently. Since we already had the single thread executor
present to ensure correct synchronization of open/file/push commands, we
are simply adding edit commands to the list.

Ideally we want to have a specialized executor that executes tasks
within the same group sequentially but groups of tasks can be executed
in parallel, thus ensuring sufficient throughput. The latter will take
much longer and will require significant rewrite of the command
execution.

Related to #8770.

JaroslavTulach reviewed (approved):
Synchronous execution is more predictable than asynchronous one and less likely to show unexpected variations.

hubertp committed:
Attempt at providing EditFileCmd implementation for tests

hubertp committed:
preprocessor requires Java

hubertp committed:
Make Handler lookup-able

The code can now properly mock the origianl Handler by looking up the
service, if avaialble.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/Handler.scala:
```diff
@@ -145,4 +82,10 @@ final class Handler {
+  def cmdFactory: CommandFactory
+}
+
+class HandlerImpl extends Handler {
```
The default implementation should be `private`, not visible to outside world.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/HandlerFactoryImpl.scala:
```diff
@@ -0,0 +1,5 @@
+package org.enso.interpreter.instrument
+
+object HandlerFactoryImpl extends HandlerFactory {
```
Again, this should ideally be `private`.

JaroslavTulach commented on engine/runtime-instrument-common/src/test/scala/org/enso/interpreter/instrument/command/SlowEditFileCmd.scala:
```diff
@@ -0,0 +1,20 @@
+    ec: ExecutionContext
+  ): Unit = {
+    if (counter % 2 == 1) {
+      Thread.sleep(2000)
```
Nice, but it will slow test execution by 2s, at least.

hubertp committed:
A test demonstrating stability issues

Previously the test would fail immediately, illustrating the underlying
problme.
Another thing to consider is to introduce some non-determinism by
removing sequential execution of commands which appears to be present in
most of runtime tests.

hubertp committed:
PR review

hubertp committed:
fix native image build
